mitodl/ocw-studio
# Github backend and API wrapper

#### Pre-Flight checklist


#### What are the relevant tickets?
#209 
Also closes #213

#### What's this PR do?
- Creates a github-specific backend and API for creating website repos and syncing website content between the repo and database.
- Creates an `updated_by` field to track who last edited a `WebsiteContent` object (which will be passed on to git, unless `FEATURE_GIT_ANONYMOUS_COMMITS=False` in yoru .env file).
- Modifies `WebsiteContent` serializers to pass on the request user to the `owner` field (if created) or `updated_by` field (if edited)

#### How should this be manually tested?
Tests should pass.

For manual testing:

In github: 
  - create a git token (Settings/Developer settings/Personal access tokens) with the following scopes: `repo`, `admin:org`, `admin: repo_hook`, `admin:org_hook` - and set `GIT_TOKEN=<your_token>` in your .env file
  - create a new organization and set `GIT_ORGANIZATION=<your_organization>` in your .env file
  - set `CONTENT_SYNC_BACKEND=content_sync.backends.github.GithubBackend` in your env file

In ocw-studio:
- Create a new website through the front-end.  Very shortly you should see a new `website.name` git repo under your organization.
- If you create new `WebsiteContent` objects via the app, they will not be synced to github because `content_filepath` is `None`.  However, the `owner` and `updated_by` fields should now be set to your user when creating it, and the `updated_by` field should be set to the user who edits it.


In a django shell, you can manually assign a fake filepath to some of these objects and then they should be added to the repo:

```python
from content_sync.tasks import *
from websites.models import *
website = Website.objects.get(name="<my_new_website>")
content = WebsiteContent.objects.create(website=website, content_filepath="content/my_content.md", markdown="Test")
sync_website_content.delay(website.name)
```

Then you can try deleting it:
```python
from content_sync.backends.github import GithubBackend
backend = GithubBackend(website)
backend.delete_content_in_backend(content.content_sync_state)
```

If you have imported any OCW courses, try syncing all of its content in 1 commit:
```python
website = Website.objects.get(name="6-253-convex-analysis-and-optimization-spring-2012")
backend = GithubBackend(website)
backend.create_website_in_backend()
backend.sync_all_content_to_backend()
```

Run ```backend.sync_all_content_to_backend()``` again no new commit should be made because there have been no changes.

Make a change to one `WebContent` file.  Assign a user to `updated_by` field as well if there isn't one yet. This time there should be another commit made with that one file, made by that user, after you save the content.

```python
content = website.websitecontent_set.last()
content.updated_by = <your_user>
content.markdown = content.markdown + " edit"
content.save()
sync_website_content.delay(website.name)
```

Merge the main branch to the preview branch and confirm the merge commit is in github:
```python
backend.create_backend_preview()
```

Then merge the preview branch to the release branch:
```python
backend.create_backend_release()
```

#### Any background context you want to provide?
The code to transform WebsiteContent data into the appropriate file content, and vice versa, is out of scope yet required for this PR to work, so some basic functionality was added as a temporary solution.

Also not included in this PR:
  - a signal to delete content from the repo when it is deleted from the database
  - generating repos for existing websites via management command
  - other management commands to sync between the database and repo(s).
